### PR TITLE
[jenkins] - add support for building binary addons

### PR DESCRIPTION
**Pull request by Memphiz**
This basically allows us to build the binary addons on jenkins. Also for normal kodi builds they won't built anymore as part of depends step, but as seperate step (which is also documented in the docs).

This is some ground work for for making use of the new binary-addons matrix job on jenkins (which will start working properly as soon as the binary_addons_pvr branch from Montellese is ready for prime time).

I have done some basic build tests and am confident that i didn't break normal kodi building. Though as so often it will show any problems once we had some tree and repo switches on jenkins due to PR building and we should expect murphy for sure.

Yes  there will be a trigger "jenkins build addons please" at one point ;)

As usual i have tried to optimize the builds a bit on the bash compatible platforms. Building bin addons does:
1. Whenever there is a change in tools/depends it will rebuild tools/depends/native and addons. 
2. Whenever there is a change in projects/cmake it will rebuild addons. 
3. If neither of both are changed - the build will be a green nop (only checkout, run through, green).

### CODE REVIEW

**Comment by Memphiz** [comment]
@wsnipex any last famous words before i let jenkins merge this?

**Comment by Montellese** [comment]
I wouldn't include `xbmc-pvr-addons` in this PR yet as it isn't part of binary addons yet.

I will owe you a lot of beer for this.

**Comment by Memphiz** [comment]
I have added xbmc-pvr-addons because the depends call isn't altered (just that in master there is still the non cmake buildsys behind it) - i didn't want to do the changes again and wanted to be consistent. AudioEncoder and PVR are already binary addons in mainline - just both with a different buildsys. The binary-addon job won't work with mainline - but that doesn't matter because it will start working when your work is done.

tl;dr; no harm done by doing it already for xbmc-pvr-addons - removing it from this PR now is a bit too much pita and unneeded ;)

**Comment by notspiff** [comment]
Please do keep single addon builds in mind. Think ahead to when they are actual addons and a pull is to be processed. This is basically allowing passing addons_to_build or whatever the cmake var was named plus an additional prestep which generates the appropriate .txt file

**Comment by Memphiz** [comment]
out of the scope of this pr - i did this basically so that montellese can rebase ontop of it and then do the further development with the help of jenkins (which reduces buildtime to 5-10min from 10-60min now). I know that this is not the finaly jenkins job we will have. Its just the start.

**Comment by Memphiz** [comment]
Beside what you said there are alot of other todos aswell (like packaging and uploading) - i am aware of that - baby steps for now...

**Comment by notspiff** [comment]
Packaging is all done.

**Comment by Memphiz** [comment]
i am speaking of the jenkins job - i didn't hook it up yet...

**Comment by Memphiz** [comment]
Correction - the binary-addons jenkins job also works for current mainline (cause the old pvr buildsys also handles depends):

http://jenkins.kodi.tv/view/Addons/job/BINARY-ADDONS/19/

**Comment by notspiff** [comment]
Well in any case; future iterations be aware that all addons have an 'addon-package' target which creates a zip ready for injection into a repo (or used through install from zip).

**Comment by Memphiz** [comment]
I remember that from ancient times when i fiddled with it already :) - need to think how to integrate all this (consider the job this PR adds support for as some sort of buildtest job for now).

**Comment by notspiff** [comment]
https://github.com/notspiff/audiodecoder-addons outlines my ideas for this part (dont mind the simpler format). A repo is the collection of addons definitions. A pull will come in to this type of repo. Saves all devs having to fork all sources, pr can easily be created through hub webif.

**Comment by wsnipex** [comment]
thanks! that will make build testing a lot faster. Adding in packaging later on shouldn't be difficult.

I'm wondering if we shouldn't just symlink the jenkins build steps to a (few) common files, since they are mostly the same instead of duplicating? Could be something for the next iteration though.

**Comment by FernetMenta** [comment]
@Montellese is it ok if we continue on opdenkamp repo until this is finished, then do the move in one step? I want to merge some fixes and backport to Helix. in addition we have two api changes queued up.

**Comment by Montellese** [comment]
@FernetMenta: go ahead as I don't want to block PVR development. At some point when everything is ready we'll have to stop, sync and move the git repositories.

**Comment by Memphiz** [comment]
mhh could we close this pr and open it again - the comment list is so long that it stalls the browser

**Comment by Memphiz** [comment]
err i ment the binary addon pr - sorry

**Comment by Memphiz** [comment]
@wsnipex - the way the builsteps are now prevent any de-duplication sadly - just try it out locally and you will fail on every idea you might have ;) (XBMC_PLATFORM_DIR is the culprit)

**Comment by wsnipex** [comment]
@Memphiz just to make sure: unified depends must be at least configured for the addons to work.
The reason is that we use config.sub|guess|site from autotools depends.

and no prob with the duplication. maybe some day we find a nice way.

**Comment by Memphiz** [comment]
yes its configured - we also build tools/depends/native before the binary addons can be built. The job calls prepare-depends, configure-depends, make-binary-addons (which internally calls make-native-depends)

**Comment by Memphiz** [comment]
Wtf i rebased this morning and gray again ...

**Comment by wsnipex** [comment]
lets get this in. jenkins build this please

**Comment by Memphiz** [comment]
failed on linux64 because of

CMake Error at CMakeLists.txt:185 (message):
  ERROR downloading
  https://github.com/xbmc/audioencoder.flac/archive/62c2cc8.tar.gz - status:
  7;"Couldn't connect to server" log: About to connect() to github.com port
  443 (#0)

And gray the button is again ... what is it with changing the pvr-addons-revision 3 times on one day @fernetmenta @opdenkamp - can't we queue this up a bit?

**Comment by Memphiz** [comment]
rebased - kicked the pr only on linux64 again to be sure:

http://jenkins.kodi.tv/job/LINUX-64/3238/

**Comment by Memphiz** [comment]
jenkins build addons please - lets see if this works too ;)
